// @darraghor/eslint-plugin-nestjs-typed@5.2.1/dist/rules/providerInjectedShouldMatchFactory/ProviderInjectedShouldMatchFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasMismatchedInjected = void 0;
const createRule_1 = require("../../utils/createRule");
// eslint-disable-next-line unicorn/import-style
//import util from "util";
const nestProviderAstParser_1 = require("../../utils/nestModules/nestProviderAstParser");
const hasMismatchedInjected = (node) => {
    // should be a nest provider - note this doesn't check the Provider used is an import actually from nest. Assumes nest Provider. Will change if this is annoying:)
    // edit 03/06/2023 - it was annoying and someone complained on github so I added a check for a "useFactory" property on the Provider declaration
    const isNestProvider = node.id.typeAnnotation
        ?.typeAnnotation
    // prettier-ignore
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ?.typeName?.name === "Provider" &&
        // and there is a useFactory property in the declaration
        nestProviderAstParser_1.nestProviderAstParser.findProvideProperty(node, "useFactory");
    if (!isNestProvider) {
        return false;
    }
    // count number of factory params
    const factoryParameterCount = nestProviderAstParser_1.nestProviderAstParser.findProvideProperty(node, "useFactory")
        ?.value?.params?.length;
    // Count number of injected params
    const injectedParameter = nestProviderAstParser_1.nestProviderAstParser.findProvideProperty(node, "inject")?.value;
    const injectedParameterCount = injectedParameter
        ? injectedParameter.elements.length
        : 0;
    // is there a mismatch?
    return injectedParameterCount !== factoryParameterCount;
};
exports.hasMismatchedInjected = hasMismatchedInjected;
const rule = (0, createRule_1.createRule)({
    name: "provided-injected-should-match-factory-parameters",
    meta: {
        docs: {
            description: "The injected items in a provider should typically match the parameters to the factory method used",
        },
        messages: {
            mainMessage: `The injected items don't match the factory method parameters, did you forget to add one?`,
        },
        schema: [],
        type: "problem",
    },
    defaultOptions: [],
    create(context) {
        return {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            VariableDeclarator(node) {
                if ((0, exports.hasMismatchedInjected)(node)) {
                    context.report({
                        node: node,
                        messageId: "mainMessage",
                    });
                }
            },
        };
    },
});
exports.default = rule;
