// @darraghor/eslint-plugin-nestjs-typed@5.2.1/dist/rules/apiPropertyMatchesPropertyOptionality/apiPropertyMatchesPropertyOptionality.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldUseOptionalDecorator = exports.shouldUseRequiredDecorator = void 0;
const createRule_1 = require("../../utils/createRule");
const typedTokenHelpers_1 = require("../../utils/typedTokenHelpers");
const shouldUseRequiredDecorator = (node) => {
    const hasOptionalDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, ["ApiPropertyOptional"]);
    const isOptionalPropertyValue = typedTokenHelpers_1.typedTokenHelpers.isOptionalPropertyValue(node);
    return hasOptionalDecorator && !isOptionalPropertyValue;
};
exports.shouldUseRequiredDecorator = shouldUseRequiredDecorator;
const shouldUseOptionalDecorator = (node) => {
    const hasRequiredDecorator = typedTokenHelpers_1.typedTokenHelpers.nodeHasDecoratorsNamed(node, ["ApiProperty"]);
    const isOptionalPropertyValue = typedTokenHelpers_1.typedTokenHelpers.isOptionalPropertyValue(node);
    return hasRequiredDecorator && isOptionalPropertyValue;
};
exports.shouldUseOptionalDecorator = shouldUseOptionalDecorator;
const rule = (0, createRule_1.createRule)({
    name: "api-property-matches-property-optionality",
    meta: {
        docs: {
            description: "Properties should have correct @ApiProperty decorators",
        },
        messages: {
            shouldUseOptionalDecorator: `Property marked as optional should use @ApiPropertyOptional decorator`,
            shouldUseRequiredDecorator: `Property marked as required should use @ApiProperty decorator`,
        },
        schema: [],
        hasSuggestions: false,
        type: "suggestion",
    },
    defaultOptions: [],
    create(context) {
        return {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            PropertyDefinition(node) {
                if ((0, exports.shouldUseOptionalDecorator)(node)) {
                    context.report({
                        node: node,
                        messageId: "shouldUseOptionalDecorator",
                    });
                }
                if ((0, exports.shouldUseRequiredDecorator)(node)) {
                    context.report({
                        node: node,
                        messageId: "shouldUseRequiredDecorator",
                    });
                }
            },
        };
    },
});
exports.default = rule;
