// @darraghor/eslint-plugin-nestjs-typed@5.2.1/dist/rules/sortModuleMetadataArrays/sortModuleMetadataArrays.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRelevantNodeName = exports.isValidModuleMetaPropertyType = void 0;
const utils_1 = require("@typescript-eslint/utils");
const createRule_1 = require("../../utils/createRule");
const wellKnownSelectors_1 = require("../../utils/wellKnownSelectors");
const utils_2 = require("@typescript-eslint/utils");
// Inspired by https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/src/rules/sort-ngmodule-metadata-arrays.ts
const DEFAULT_LOCALE = "en-US";
const isValidModuleMetaPropertyType = (node) => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return (!!node &&
        utils_2.ASTUtils.isNodeOfTypes([
            utils_1.TSESTree.AST_NODE_TYPES.Identifier,
            utils_1.TSESTree.AST_NODE_TYPES.CallExpression,
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ])); // unsure about this type
};
exports.isValidModuleMetaPropertyType = isValidModuleMetaPropertyType;
const getRelevantNodeName = (node) => {
    let currentName = "";
    if (node.type === utils_1.TSESTree.AST_NODE_TYPES.Identifier) {
        currentName = node.name;
    }
    if (node.type === utils_1.TSESTree.AST_NODE_TYPES.CallExpression &&
        node.callee.type === utils_1.TSESTree.AST_NODE_TYPES.MemberExpression &&
        node.callee.object.type === utils_1.TSESTree.AST_NODE_TYPES.Identifier) {
        currentName = node.callee.object.name;
    }
    return currentName;
};
exports.getRelevantNodeName = getRelevantNodeName;
const defaultLocaleOptions = [
    {
        locale: DEFAULT_LOCALE,
    },
];
exports.default = (0, createRule_1.createRule)({
    name: "sort-module-metadata-arrays",
    meta: {
        type: "suggestion",
        docs: {
            description: "Ensures ASC alphabetical order for `Module` metadata arrays for easy visual scanning",
        },
        fixable: "code",
        schema: [
            {
                type: "object",
                properties: {
                    locale: {
                        type: "string",
                        description: "A string with a BCP 47 language tag.",
                        default: DEFAULT_LOCALE,
                    },
                },
                additionalProperties: false,
            },
        ],
        messages: {
            moduleMetadataArraysAreSorted: "`Module` metadata arrays should be sorted in ASC alphabetical order",
        },
    },
    defaultOptions: defaultLocaleOptions,
    create(contextWithoutDefaults) {
        const context = contextWithoutDefaults.options &&
            contextWithoutDefaults.options.length > 0
            ? contextWithoutDefaults
            : // only apply the defaults when the user provides no config
                Object.setPrototypeOf({
                    options: defaultLocaleOptions,
                }, contextWithoutDefaults);
        const { locale } = context.options[0];
        const sourceCode = context.getSourceCode();
        return {
            [`${wellKnownSelectors_1.MODULE_CLASS_DECORATOR} Property > ArrayExpression`]({ elements, }) {
                const unorderedNodes = elements
                    // nestjs modules use identifiers and call expressions
                    // can modify this later
                    .filter(exports.isValidModuleMetaPropertyType)
                    .map((current, index, list) => [current, list[index + 1]])
                    .find(([current, next]) => {
                    return (current &&
                        next &&
                        (0, exports.getRelevantNodeName)(current).localeCompare((0, exports.getRelevantNodeName)(next), locale) === 1);
                });
                if (!unorderedNodes)
                    return;
                const [unorderedNode, nextNode] = unorderedNodes;
                context.report({
                    node: nextNode, // can't really be null
                    messageId: "moduleMetadataArraysAreSorted",
                    fix: (fixer) => [
                        fixer.replaceText(unorderedNode, sourceCode.getText(nextNode)),
                        fixer.replaceText(nextNode, sourceCode.getText(unorderedNode)),
                    ],
                });
            },
        };
    },
});
